### PR TITLE
Grade distribution parser

### PR DESCRIPTION
**Pull request by phongfrancis**
Complete #31.

** Only parses 1 term for now. If you want to complete every term file in there, change '**parseTermFile**' function to '**parseAllFiles**' and pass in the correct path. I tested it and it takes approximately **55** seconds to parse all term files. **

### CODE REVIEW

File: src/main/java/com/pscheduler/util/parser/GradeParser.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+                .B(Double.parseDouble(values[9]))
+                .C(Double.parseDouble(values[10]))
+                .D(Double.parseDouble(values[11]))
+                .F(Double.parseDouble(values[12]))
+                .withdraws(Integer.parseInt(values[13]))
+                .term(termID);
+
+        return gradeBuilderFactory.build();
+    }
+
```
Comment (PhillipNgo) [inline src/main/java/com/pscheduler/util/parser/GradeParser.java]: Extra newline

File: src/main/java/com/pscheduler/util/GradeBuilderFactory.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
+package com.pscheduler.util;
+
+import com.pscheduler.server.model.CourseGPA;
+
+public class GradeBuilderFactory {
```
Comment (PhillipNgo) [inline src/main/java/com/pscheduler/util/GradeBuilderFactory.java]: Technically just a builder right now. If we get to the serverless part might need to add the factory part.

File: src/main/java/com/pscheduler/server/core/DatabaseLoader.java [hunk 3]
```diff
@@ -45,7 +48,17 @@ public void run(ApplicationArguments args) throws Exception {
         }
         courses.save(courseList);
 
-        CourseGPA course = new CourseGPA(
+
+        GradeParser grader = new GradeParser();
+
+        // REMEMBER TO CHANGE THIS TO PARSEALLFILES
```
Comment (PhillipNgo) [inline src/main/java/com/pscheduler/server/core/DatabaseLoader.java]: Remove this comment for now, we can change it later

File: src/main/java/com/pscheduler/server/core/DatabaseLoader.java [hunk 4]
```diff
@@ -45,7 +48,17 @@ public void run(ApplicationArguments args) throws Exception {
         }
         courses.save(courseList);
 
-        CourseGPA course = new CourseGPA(
+
```
Comment (PhillipNgo) [inline src/main/java/com/pscheduler/server/core/DatabaseLoader.java]: extra newline

File: src/main/java/com/pscheduler/server/core/DatabaseLoader.java [hunk 5]
```diff
@@ -45,7 +48,17 @@ public void run(ApplicationArguments args) throws Exception {
...
         courses.save(courseList);
 
-        CourseGPA course = new CourseGPA(
+
+        GradeParser grader = new GradeParser();
+
+        // REMEMBER TO CHANGE THIS TO PARSEALLFILES
+        List<CourseGPA> courseGrades = grader.parseTermFile(gpaDataPath + "/2019/Winter2019.csv");
+        for (CourseGPA grade : courseGrades) gpa.save(grade);
+
```
Comment (PhillipNgo) [inline src/main/java/com/pscheduler/server/core/DatabaseLoader.java]: a lot of new lines?

File: src/main/java/com/pscheduler/server/core/DatabaseLoader.java [hunk 6]
```diff
@@ -45,7 +48,17 @@ public void run(ApplicationArguments args) throws Exception {
...
+
+
+
+        /*CourseGPA course = new CourseGPA(
```
[the remark this answers is not in the document]
Comment (phongfrancis) [reply src/main/java/com/pscheduler/server/core/DatabaseLoader.java]: Removed..

File: src/main/java/com/pscheduler/server/model/CourseGPA.java [hunk 7]
```diff
@@ -220,4 +227,8 @@ public int getWithdraws() {
     public void setWithdraws(int withdraws) {
         this.withdraws = withdraws;
     }
+
+    public int getTerm() { return this.term; }
```
[the remark this answers is not in the document]
Comment (phongfrancis) [reply src/main/java/com/pscheduler/server/model/CourseGPA.java]: Fixed.

File: src/main/java/com/pscheduler/server/model/CourseGPA.java [hunk 8]
```diff
@@ -220,4 +227,8 @@ public int getWithdraws() {
...
+
+    public void setTerm(int term) { this.term = term; }
```
[the remark this answers is not in the document]
Comment (phongfrancis) [reply src/main/java/com/pscheduler/server/model/CourseGPA.java]: Fixed.

File: src/main/java/com/pscheduler/server/repository/RepositoryConfig.java [hunk 9]
```diff
@@ -10,5 +11,6 @@
     @Override
     public void configureRepositoryRestConfiguration(RepositoryRestConfiguration config) {
         config.exposeIdsFor(Course.class);
+        config.exposeIdsFor(CourseGPA.class);
     }
 }
```
[the remark this answers is not in the document]
Comment (phongfrancis) [reply src/main/java/com/pscheduler/server/repository/RepositoryConfig.java]: Fixed.

File: src/main/java/com/pscheduler/util/GradeBuilderFactory.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+              A,
+              B,
+              C,
+              D,
+              F,
+              withdraws,
+              term
+        );
+    }
+
```
[the remark this answers is not in the document]
Comment (phongfrancis) [reply src/main/java/com/pscheduler/util/GradeBuilderFactory.java]: Fixed.

File: src/main/java/com/pscheduler/util/parser/GradeParser.java (hunk 1 above)
Comment (phongfrancis) [reply src/main/java/com/pscheduler/util/parser/GradeParser.java]: Removed

File: src/main/java/com/pscheduler/util/GradeBuilderFactory.java (hunk 2 above)
Comment (phongfrancis) [reply src/main/java/com/pscheduler/util/GradeBuilderFactory.java]: Renamed to just 'GradeBuilder' now.

File: src/main/java/com/pscheduler/server/core/DatabaseLoader.java (hunk 3 above)
Comment (phongfrancis) [reply src/main/java/com/pscheduler/server/core/DatabaseLoader.java]: Removed

File: src/main/java/com/pscheduler/server/core/DatabaseLoader.java (hunk 4 above)
Comment (phongfrancis) [reply src/main/java/com/pscheduler/server/core/DatabaseLoader.java]: Removed.

File: src/main/java/com/pscheduler/server/core/DatabaseLoader.java (hunk 5 above)
Comment (phongfrancis) [reply src/main/java/com/pscheduler/server/core/DatabaseLoader.java]: Gone.

File: src/main/java/com/pscheduler/util/parser/GradeParser.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+     * Gets the term from a given file name
+     *
+     * @param term the name of the file
+     * @return the term ID
+     */
+    public int getTerm(String term) {
+        String semester = term.substring(0, term.length() - 4);
+        int month = 0;
+
+        if (semester.equals("Spring")) {
```
[the remark this answers is not in the document]
Comment (phongfrancis) [reply src/main/java/com/pscheduler/util/parser/GradeParser.java]: Done.

File: src/main/java/com/pscheduler/util/parser/GradeParser.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+     * @return the parsed course
+     */
+    private CourseGPA makeClass(String line, String term) {
+        int termID = getTerm(term);
+
+        gradeBuilderFactory.reset();
+        String[] values = line.split(SPLIT_DELIMETER);
+
+        if (values.length != 14) {
+            System.out.println("bad line formatting: " + line);
```
[the remark this answers is not in the document]
Comment (phongfrancis) [reply src/main/java/com/pscheduler/util/parser/GradeParser.java]: Never does. Removing it.

File: src/main/java/com/pscheduler/util/parser/GradeParser.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (phongfrancis) [reply src/main/java/com/pscheduler/util/parser/GradeParser.java]: Done.

File: src/main/java/com/pscheduler/util/parser/GradeParser.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (phongfrancis) [reply src/main/java/com/pscheduler/util/parser/GradeParser.java]: It actually does. The end of each csv file has an empty line for some reason. So it gets called on every file with a spliced line (using comma as delimiter) on the very last row. So it is necessary for now until we can find a way around this issue.

File: src/main/java/com/pscheduler/util/parser/GradeParser.java [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+     * @param path path to folder containing term year (which contains term files)
+     * @return list holding the gpa data for each class
+     * @throws Exception
+     */
+    public List<CourseGPA> parseAllFiles(String path) throws Exception {
+        List<CourseGPA> courses = new ArrayList<>();
+        File folder = new File(path);
+        File[] files = folder.listFiles();
+
+        for (int x = 0; x < files.length; x++) {
```
[the remark this answers is not in the document]
Comment (phongfrancis) [reply src/main/java/com/pscheduler/util/parser/GradeParser.java]: Done.

File: src/main/java/com/pscheduler/util/parser/GradeParser.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (phongfrancis) [reply src/main/java/com/pscheduler/util/parser/GradeParser.java]: Fixed to check if the line length is 1

File: src/main/java/com/pscheduler/util/parser/GradeParser.java [hunk 14]
```diff
@@ -0,0 +1,151 @@
+package com.pscheduler.util.parser;
+
```
Comment (PhillipNgo) [inline src/main/java/com/pscheduler/util/parser/GradeParser.java]: extra newline

**Comment by PhillipNgo** [comment]
@phongfrancis Take a look at my changes in 7fe575885c979fb858daf9733d08234f12cbaa4b. Let me know if any of it doesn't make sense, then I'll merge it.

**Pull request merged**